### PR TITLE
Bonsai Tries Rolling Head and Layered Reads

### PR DESCRIPTION
**Pull request by shemnon**
Two updates to Bonsai Tries

* Log Rolling is implemented on top of the existing Persisted head.  When Besu is at chain head and the new best head makes the current head an orphan branch, the Bonsai TrieLogs are used to roll back to a common block and roll forward to the needed base block.  Goerli is known to maintain sync.  There are still some issues with frontier era block receipts.
* Non-mutable reads can be done off of the persisted block.  These are accurate for all reads that were performed in the block.  If a read is not known it proceeds through a fallback series of calls to prior layers until it hits the persisted block.  These layered reads are driven off of the TrieLogs.

## Changelog

- [ ] I thought about the changelog and included a [changelog update if required](https://wiki.hyperledger.org/display/BESU/Changelog).

### CODE REVIEW

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/BonsaiLayeredWorldState.java [hunk 1]
```diff
@@ -33,21 +32,73 @@
...
+      }
+      if (currentLayer.parent == null) {
+        currentLayer = null;
+      } else if (currentLayer.parent instanceof BonsaiLayeredWorldState) {
+        currentLayer = (BonsaiLayeredWorldState) currentLayer.parent;
+      } else {
+        return currentLayer.parent.getCode(address);
+      }
+    }
+    return null;
```
Comment (matkt) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/BonsaiLayeredWorldState.java:84]: Should we not return an OptionalEmpty instead ? Or did you do that to not impact Forest ?

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/BonsaiLayeredWorldState.java [hunk 2]
```diff
@@ -33,21 +32,73 @@
...
+    return height;
+  }
+
   @Override
   public Bytes getCode(final Address address) {
-    return trieLog.getCode(address).orElseGet(() -> parent.getCode(address));
+    // this must be iterative and lambda light because the stack may blow up
+    // mainly because we don't have tail calls.
+    BonsaiLayeredWorldState currentLayer = this;
+    while (currentLayer != null) {
```
Comment (matkt) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/BonsaiLayeredWorldState.java:71]:
```
 while (currentLayer != null) {
      final Optional<Bytes> maybeCode = currentLayer.trieLog.getCode(address);
      if (maybeCode.isPresent()) {
        return maybeCode.get();
      }
      if (currentLayer.parent == null) {
        currentLayer = null;
      } else if (currentLayer.parent instanceof BonsaiLayeredWorldState) {
        currentLayer = (BonsaiLayeredWorldState) currentLayer.parent;
      } else {
        return currentLayer.parent.getCode(address);
      }
    }
```

NIT : I have seen this logic to go from layer to another layer in several places in the code. Is it interesting to make this part generic ? and maybe use lambdas for what is not generic (getCode, etc.) ?

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/chain/GenesisState.java [hunk 3]
```diff
@@ -102,13 +102,13 @@ public Block getBlock() {
...
   public void writeStateTo(final MutableWorldState target) {
-    writeAccountsTo(target, genesisAccounts, (Hash) block.getHeader().getBlockHash());
+    writeAccountsTo(target, genesisAccounts, block.getHeader());
   }
 
   private static void writeAccountsTo(
       final MutableWorldState target,
       final List<GenesisAccount> genesisAccounts,
-      final Hash rootHash) {
+      final BlockHeader rootHash) {
```
Comment (matkt) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/chain/GenesisState.java:111]: shouldn't we rename this rootHash  to blockHeader ?

**Review by matkt — APPROVED** [review]
LGTM. Some comments . just looked at the code

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/BonsaiLayeredWorldState.java (hunk 2 above)
Comment (shemnon) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/BonsaiLayeredWorldState.java:71]: We would need a lambda to detect a hit (optional:isPresent or != null) and the lambda to fetch the value.  It looks verbose but I'm concerned wrapping it in a method would negatively impact performance.  The initial recursive approach resulted in stack overflow errors, so a in-place iterative search is my preference.

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/BonsaiLayeredWorldState.java (hunk 1 above)
Comment (shemnon) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/BonsaiLayeredWorldState.java:84]: done.  It doesn't escape the bonsai classes.

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/chain/GenesisState.java (hunk 3 above)
Comment (shemnon) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/chain/GenesisState.java]: done.

**Pull request merged**